updater/src/updater.rs

```diff
@@ -255,47 +380,151 @@ impl Updater {
+											state.status = UpdaterStatus::Disabled;
+											warn!("{}", err);
+										} else {
+											state.status = UpdaterStatus::Ready { release: release.clone() };
+											// will lock self.state
+											drop(state);
+											this.updater_step();
+										}
+									},
+									// There was an error fetching the update, apply a backoff delay before retrying
+									Err(err) => {
+										let delay = 2usize.pow(retries) as u64;
+										let backoff = (retries, Instant::now() + Duration::from_secs(delay));
+
+										state.status = UpdaterStatus::FetchBackoff { release: release.clone(), backoff, binary };
+
+										warn!("Unable to fetch update ({}): {:?}, retrying in {} seconds.", release.version, err, delay);
+									},
+								}
+							}
+						}
+					}
 				};
-				s.ready = Some(fetched);
-				auto
+
+				self.fetcher.fetch(binary, Box::new(on_fetch));
 			};
-			if auto {
-				// will lock self.state, so ensure it's outside of previous block.
-				self.execute_upgrade();
+
+			match state.status.clone() {
+				// updater is disabled
+				UpdaterStatus::Disabled => {},
+				// the update has already been installed
+				UpdaterStatus::Installed { ref release, .. } if *release == latest.track => {},
+				// we're currently fetching this update
+				UpdaterStatus::Fetching { ref release, .. } if *release == latest.track => {},
+				// the fetch has failed and we're backing off the next retry
+				UpdaterStatus::FetchBackoff { ref release, backoff, .. } if *release == latest.track && Instant::now() < backoff.1 => {},
+				// we're delaying the update until the given block number
+				UpdaterStatus::Waiting { ref release, block_number, .. } if *release == latest.track && current_block_number < block_number => {},
+				// we're at (or past) the block that triggers the update, let's fetch the binary
+				UpdaterStatus::Waiting { ref release, block_number, binary } if *release == latest.track && current_block_number >= block_number => {
+					info!(target: "updater", "Update for binary {} triggered", binary);
+
+					state.status = UpdaterStatus::Fetching { release: release.clone(), binary, retries: 1 };
+					// will lock self.state
+					drop(state);
+					fetch(binary);
+				},
+				// we're ready to retry the fetch after we applied a backoff for the previous failure
+				UpdaterStatus::FetchBackoff { ref release, backoff, binary } if *release == latest.track && Instant::now() >= backoff.1 => {
+					state.status = UpdaterStatus::Fetching { release: release.clone(), binary, retries: backoff.0 + 1 };
+					// will lock self.state
+					drop(state);
+					fetch(binary);
+				},
+				// the update is ready to be installed
+				UpdaterStatus::Ready { ref release } if *release == latest.track => {
+					let auto = match self.update_policy.filter {
+						UpdateFilter::All => true,
+						UpdateFilter::Critical if release.is_critical /* TODO: or is on a bad fork */ => true,
```

Comment:
Intentionally or unintentionally left TODO?

Reply:
This was copied from the existing code. I'm not sure how to fix it.